### PR TITLE
Add Marconi runtime view in documentation

### PR DESCRIPTION
**Pull request by koslambrou**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

File: doc/read-the-docs-site/architecture/runtime-view.rst [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+The input source reader provides an interface to fetch data from any datasource: a local Cardano node, Ogmios, cardano-db-sync, etc.
+However, the input data needs to be mapped to a ``ProcessedInput``.
+The ``ProcessedInput`` is sent to the preprocessor which applies some preprocessing function on the inputs.
+The preprocessed inputs are sent to the transformer which also transforms the inputs, and sends them to the indexer.
+Finally, the indexer will index them to it's storage backend.
+
+.. note::
+
+  See the explanation of the `marconi-core` components in order to understand concepts like ``Preprocessor``,  ``Transformer``, ``Indexer``, etc..
```
Comment (ana-pantilie) [inline doc/read-the-docs-site/architecture/runtime-view.rst:47]: I realize that this isn't merged yet, would be good to provide a link after it is

File: doc/read-the-docs-site/architecture/runtime-view.rst [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    Container_Boundary(marconiContainer, "Marconi application") {
+      Component(sourceReader, "Local node source reader", "Haskell component", "Module that processes data from the local node and transforms it to a [ProcessedInput]")
+    }
+
+    Rel(blockchain, cardanoNode, "sends blocks to")
+    Rel(cardanoNode, sourceReader, "sends chain-sync event")
+
+The local Cardano node constructs a local state of the Cardano blockchain by fetching blocks from its peers and storing it on disk.
+Then, node clients can request for blocks by the using the node's chain-sync mini-protocol.
+The chain-sync protocol essentially streams events to the client which have the following structure:
```
Comment (ana-pantilie) [inline doc/read-the-docs-site/architecture/runtime-view.rst:68]: This is interesting, could you provide a link to the documentation of this protocol?

File: doc/read-the-docs-site/architecture/runtime-view.rst [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    Rel(coordinator, indexerC, "forwards input")
+    Rel(indexerD, indexerA, "queries")
+    Rel(indexerD, indexerB, "queries")
+
+    Rel(dApp, indexerC, "queries")
+    Rel(dApp, indexerD, "queries")
+
+In this view, we show that the ``Coordinator`` can combine multiple indexers.
+The ``Coordinator``'s role is to foward the incoming inputs to each of the indexer and make sure each indexer has processed the input before moving to the next one.
+Additionnaly, some indexer can be "read-only" as they won't index any information, but can provide queries that aggregate the query results of other indexers.
```
Comment (ana-pantilie) [inline doc/read-the-docs-site/architecture/runtime-view.rst:113]: Typos?
```suggestion
Additionally, some indexers can be "read-only" as they won't index any information, but can provide queries that aggregate the query results of other indexers.
```

Hmm, I don't understand the connection of an indexer being "read-only" and not indexing information. What I mean is, doesn't indexing mean by definition mean _reading_ some information and processing it?

File: doc/read-the-docs-site/architecture/runtime-view.rst [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+--------------------------
+
+In this section, we'll show the querying runtime view of a Marconi application which contains at least one indexer.
+The Marconi indexer querying is typically done in a separate thread from the indexing itself.
+We're assuming that the indexers have fully consumed the input stream and are ready to be queried.
+
+.. note::
+
+   Although you can start querying the indexer even if it did not *fully* consumed the input stream.
+   The only issue is that the user will get out of date information when compared the actual tip of the local node.
```
Comment (ana-pantilie) [inline doc/read-the-docs-site/architecture/runtime-view.rst:125]: Maybe rephrase this a bit?
```suggestion
   You can start querying the indexer even if it did not *fully* consume the input stream.
   However, you will get out of date information until it has synced with the actual current tip of the local node.
```

File: doc/read-the-docs-site/architecture/runtime-view.rst [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    | RollBackward C.ChainPoint C.ChainTip
+    deriving (Show, Functor, Generic)
+
+where ``a`` can be anything including a block.
+
+Now, let's assume that we want to run multiple indexers that process inputs at the same rate, and that a subset of indexers are queryable from the a client application.
+Running multiple indexers that process inputs at the same rate are managed by a ``Coordinator``.
+The runtime view would look like:
+
+.. uml::
```
Comment (ana-pantilie) [inline doc/read-the-docs-site/architecture/runtime-view.rst:83]: Hmm, I think this part is missing from the webpage, at least for my local website. Could you double-check that it is appearing for you?

File: doc/read-the-docs-site/architecture/runtime-view.rst [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+In this section, we'll show the querying runtime view of a Marconi application which contains at least one indexer.
+The Marconi indexer querying is typically done in a separate thread from the indexing itself.
+We're assuming that the indexers have fully consumed the input stream and are ready to be queried.
+
+.. note::
+
+   Although you can start querying the indexer even if it did not *fully* consumed the input stream.
+   The only issue is that the user will get out of date information when compared the actual tip of the local node.
+
+.. uml::
```
Comment (ana-pantilie) [inline doc/read-the-docs-site/architecture/runtime-view.rst:127]: Hmm, I think this part is missing from the webpage, at least for my local website. Could you double-check that it is appearing for you?

File: doc/read-the-docs-site/architecture/runtime-view.rst (hunk 3 above)
Comment (koslambrou) [reply doc/read-the-docs-site/architecture/runtime-view.rst]: Good point. I removed the term "read-only" and replaced it with a `QueryAggregate` component.

File: doc/read-the-docs-site/architecture/runtime-view.rst (hunk 5 above)
Comment (koslambrou) [reply doc/read-the-docs-site/architecture/runtime-view.rst:87]: It's working for me. I maybe forced push a fix without realizing that it was broken.

File: doc/read-the-docs-site/architecture/runtime-view.rst (hunk 2 above)
Comment (koslambrou) [reply doc/read-the-docs-site/architecture/runtime-view.rst:68]: Added the following note:

"
   For an overview on the different Cardano node mini-protocols, see this `Cardano Docs page <https://docs.cardano.org/explore-cardano/cardano-network/networking-protocol/>`. For a more in-depth explanation, read the `Ouroboros network specification <https://input-output-hk.github.io/ouroboros-network/pdfs/network-spec/network-spec.pdf>`_.
"

File: doc/read-the-docs-site/architecture/runtime-view.rst [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+In this view, we show that the ``Coordinator`` can combine multiple indexers.
+The ``Coordinator``'s role is to forward the incoming inputs to each of the indexer and make sure each indexer has processed the input before moving to the next one.
+
+Then, we have the querying part.
+Each indexer actually defines two things:
+
+* how to index the given input
+* how to query the indexed information
+
+Therefore, dApps will simply interact directly with the indexers when needing to query some information.
```
Comment (berewt) [inline doc/read-the-docs-site/architecture/runtime-view.rst:124]: It doesn't include the need for an `MVar` here. Do we want to exclude it from the explanations to keep them simpler?

**Review by berewt — APPROVED** [review]
Looks good to me. Just a minor question.

File: doc/read-the-docs-site/architecture/runtime-view.rst (hunk 5 above)
Comment (ana-pantilie) [reply doc/read-the-docs-site/architecture/runtime-view.rst:87]: Hmm, it's supposed to render a diagram, right? I still can't see it, but it must be something wrong with my settings

**Review by ana-pantilie — APPROVED** [review]
LGTM, although I still can't see the diagrams, but if it works for you and Nicolas then I can approve.

**Comment by ana-pantilie** [comment]
Ah, figured out why the uml diagrams aren't showing, I'm getting the following error in my terminal: `/home/ana/IOG/marconi/doc/read-the-docs-site/architecture/runtime-view.rst:38: WARNING: plantuml command 'plantuml' cannot be run`

File: doc/read-the-docs-site/architecture/runtime-view.rst (hunk 7 above)
Comment (koslambrou) [reply doc/read-the-docs-site/architecture/runtime-view.rst:124]: Good point. I added a note explaining how the `MVar` integrates the query and indexing thread together.

**Comment by koslambrou** [comment]
@ana-pantilie Weird. Did you enter a Nix shell? What command did you run to launch the docs?

**Comment by ana-pantilie** [comment]
@koslambrou Yes, I'm in the nix shell. I ran `develop-rtd-site`, should I have done something else?

**Comment by koslambrou** [comment]
@ana-pantilie Weird... I'll create a separate issue for that.
